core/src/main/scala/kafka/server/ControllerApis.scala

```diff
@@ -441,23 +441,14 @@ class ControllerApis(val requestChannel: RequestChannel,
     if (apiVersionRequest.hasUnsupportedRequestVersion) {
       requestHelper.sendResponseMaybeThrottle(request,
         requestThrottleMs => apiVersionRequest.getErrorResponse(requestThrottleMs, UNSUPPORTED_VERSION.exception))
-      CompletableFuture.completedFuture[Unit](())
     } else if (!apiVersionRequest.isValid) {
       requestHelper.sendResponseMaybeThrottle(request,
         requestThrottleMs => apiVersionRequest.getErrorResponse(requestThrottleMs, INVALID_REQUEST.exception))
-      CompletableFuture.completedFuture[Unit](())
     } else {
-      val context = new ControllerRequestContext(request.context.header.data, request.context.principal, OptionalLong.empty())
-      controller.finalizedFeatures(context).handle { (result, exception) =>
-        requestHelper.sendResponseMaybeThrottle(request, requestThrottleMs => {
-          if (exception != null) {
-            apiVersionRequest.getErrorResponse(requestThrottleMs, exception)
-          } else {
-            apiVersionManager.apiVersionResponse(requestThrottleMs, result.featureMap().asScala.toMap, result.epoch())
-          }
-        })
-      }
+      requestHelper.sendResponseMaybeThrottle(request,
+        requestThrottleMs => apiVersionManager.apiVersionResponse(requestThrottleMs))
```

Comment:
Just checking the semantics here.

ApiVersionManager now has a feature provider which is calling into FeaturePublisher. This means if we handle an ApiVersionsRequest before the controller has finished loading (or at least before FeaturePublisher loads the initial metadatA), we'll immediately return the base set of KRaft features (which is just IBP_3_0_IV1).

Previously, we asynchronously called into QuorumController to get the features from FeatureControlManager. I believe this meant we would have loaded the metadata before processing the read event.

With this change, is it possible to hit the controller during it's initial loading and get an ApiVersionsResponse that has stale data?

Reply:
I think this is inevitable, the broker also get the feature data from MetadataCache,  which may also be stale data.

Reply:
I will change this not to start the listener until we've caught up to the local HWM. That will avoid this and some other gotchas.